python/cudf_polars/cudf_polars/dsl/utils/groupby.py

```diff
@@ -0,0 +1,98 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Utilities for grouped aggregations."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from cudf_polars.dsl import ir
+from cudf_polars.dsl.utils.aggregations import apply_pre_evaluation
+from cudf_polars.dsl.utils.naming import unique_names
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+    from typing import Any
+
+    import pylibcudf as plc
+
+    from cudf_polars.dsl import expr
+    from cudf_polars.utils import config
+
+__all__ = ["rewrite_groupby"]
+
+
+def rewrite_groupby(
+    node: Any,
+    schema: dict[str, plc.DataType],
+    keys: Sequence[expr.NamedExpr],
+    aggs: Sequence[expr.NamedExpr],
+    config_options: config.ConfigOptions,
+    inp: ir.IR,
+) -> ir.IR:
+    """
+    Rewrite a groupby plan node into something we can handle.
+
+    Parameters
+    ----------
+    node
+        The polars groupby plan node.
+    schema
+        Schema of the groupby plan node.
+    keys
+        Grouping keys.
+    aggs
+        Originally requested aggregations.
+    config_options
+        Configuration options.
+    inp
+        Input plan node to the groupby.
+
+    Returns
+    -------
+    New plan node representing the grouped aggregations.
+
+    Raises
+    ------
+    NotImplementedError
+        If any of the requested aggregations are unsupported.
+
+    Notes
+    -----
+    Since libcudf can only perform grouped aggregations on columns
+    (not arbitrary expressions), the approach is to split each
+    aggregation into a pre-selection phase (evaluating expressions
+    that live within an aggregation), the aggregation phase (now
+    acting on columns only), and a post-selection phase (evaluating
+    expressions of aggregated results).
+
+    This does scheme does not permit nested aggregations, so those are
+    unsupported.
+    """
+    if len(aggs) == 0:
+        # TODO: use Distinct when the partitioned executor supports it
```

Comment:
Ref: https://github.com/rapidsai/cudf/pull/18277  - That PR is quite stale now, but a `Distinct` can easily lower to a `Distinct(Repartition(Distinct(...)))` (or a `Distinct(Shuffle(Distinct(...)))` if any columns are high-cardinality).